src/database/database-mariadb.cpp

```diff
@@ -0,0 +1,1407 @@
+		stmt->close();
+
+	} catch (sql::SQLException &e) {
+		errorstream << "[MariaDB] Error: " << + e.what() << std::endl;
+		throw DatabaseException("MariaDB Error");
+	}
+}
+
+
+void Database_MariaDB::createTable(std::string table_name, std::string table_spec) {
+	try {
+
+		std::unique_ptr<sql::Statement> stmt(conn->createStatement());
+		stmt->executeUpdate("CREATE TABLE IF NOT EXISTS " + table_name + " (" + table_spec + ") CHARACTER SET = 'utf8mb4' COLLATE = 'utf8mb4_unicode_ci' ENGINE=InnoDB");
+		stmt->close();
+
+	} catch (sql::SQLException &e) {
+		errorstream << "[MariaDB] Error: " << + e.what() << std::endl;
+		throw DatabaseException("MariaDB Error");
+	}
+}
+
+
+std::int32_t Database_MariaDB::lastInsertId() {
+	try {
+		
+		std::unique_ptr<sql::ResultSet> resLastInsertId(stmtLastInsertId->executeQuery());
+
+		resLastInsertId->next();
+		std::int32_t id(resLastInsertId->getInt(1));
+
+		resLastInsertId->close();
+		
+		return id;
+
+	} catch (sql::SQLException &e) {
+		errorstream << "[MariaDB] Error: " << + e.what() << std::endl;
+		throw DatabaseException("MariaDB Error");
+	}
+}
+
+
+std::map<std::string, std::string> Database_MariaDB::parseConnectionString(std::string input) {
+	
+	input += " ";
+	std::map<std::string, std::string> params;
+	size_t pos, last_pos = 0;
+	
+	while (std::string::npos != (pos = input.find(' ', last_pos))) {
+
+		std::string part = input.substr(last_pos, pos - last_pos);
+		std::size_t pos2 = part.find('=');
+		
+		if (std::string::npos != pos2) {
+
+			params[part.substr(0, pos2)] = part.substr(pos2 + 1);
+		
+		}
+		
+		last_pos = pos + 1;
+
+	}
+
+	return params;
+}
+
+
+sql::PreparedStatement* Database_MariaDB::prepareStatement(const std::string query) {
+	try {
+
+		sql::PreparedStatement* stmt(conn->prepareStatement(query));
+
+		return stmt;
+
+	} catch (sql::SQLException &e) {
+		errorstream << "[MariaDB] Error: " << + e.what() << std::endl;
+		throw DatabaseException("MariaDB Error");
+	}
+}
+
+
+void Database_MariaDB::rollback() {
+	try {
+
+		stmtRollback->execute();
+	
+	} catch (sql::SQLException &e) {
+		errorstream << "[MariaDB] Error: " << + e.what() << std::endl;
+		throw DatabaseException("MariaDB Error");
+	}
+}
+
+
+/**
+ * 
+ * 
+ * MapDatabaseMariaDB -- The MariaDB Map Database Backend
+ * 
+ * 
+ */
+
+
+MapDatabaseMariaDB::MapDatabaseMariaDB(const std::string &connect_string) : Database_MariaDB(connect_string, ""), MapDatabase() {
+
+	// connect to database
+	connect();
+
+	// create table if it does not exist
+	createTable(
+		"blocks",
+		"x INT, "
+		"y INT, "
+		"z INT, "
+		"data BLOB, "
+		"PRIMARY KEY (x, y, z)"
+	);
+
+	try {
+
+		// prepare statements
+		stmtDeleteBlock.reset(prepareStatement(
+			"DELETE FROM blocks WHERE x = ? and y = ? and z = ?"
+		));
+
+		stmtLoadBlock.reset(prepareStatement(
+			"SELECT * FROM blocks WHERE x = ? and y = ? and z = ?"
```

Comment:
never do a SELECT *, always use precise fields

Reply:
Looking at loadBlock below, only `data` is needed

Reply:
> never do a SELECT *, always use precise fields

Why? To make changing the schema less likely to have side effects?

Reply:
Fixed in latest commit.